src/blocks/keyboard_layout/xkb_event.rs

```diff
@@ -0,0 +1,147 @@
+use super::*;
+use x11rb_async::{
+    connection::{Connection as _, RequestConnection as _},
+    protocol::{
+        xkb::{
+            self, ConnectionExt as _, EventType, MapPart, NameDetail, SelectEventsAux,
+            UseExtensionReply, ID,
+        },
+        xproto::ConnectionExt as _,
+        Event,
+    },
+    rust_connection::RustConnection,
+};
+
+const XCB_XKB_MINOR_VERSION: u16 = 0;
+const XCB_XKB_MAJOR_VERSION: u16 = 1;
+
+pub(super) struct XkbEvent {
+    connection: RustConnection,
+    cur_layout: String,
+}
+
+fn parse_layout(buf: Vec<u8>, index: usize) -> Result<String> {
+    let prefix = buf
+        .into_iter()
+        .take_while(|&c| c != b':')
+        .collect::<Vec<u8>>();
+
+    let layout = prefix
+        .split(|&c| c == b'+')
+        .skip(1) // layout names start from index 1
+        .nth(index)
+        .error("Index out of range")?
+        .to_vec();
+
+    String::from_utf8(layout).map_err(|e| Error::new(e.to_string()))
+}
```

Comment:
A non-allocating version:

```suggestion
fn parse_layout(buf: &[u8], index: usize) -> Result<&str> {
    let colon_i = buf.iter().position(|c| *c == b':').unwrap_or(buf.len());

    let layout = buf[..colon_i]
        .split(|&c| c == b'+')
        .skip(1) // layout names start from index 1
        .nth(index)
        .context("Index out of range")?;

    std::str::from_utf8(layout).context("non utf8 layout")
}
```

Reply:
perfect! applied the changes.